exporter/key_groups.go

```diff
@@ -0,0 +1,240 @@
+package exporter
+
+import (
+	"encoding/csv"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+)
+
+type keyGroupMetrics struct {
+	keyGroup    string
+	count       int64
+	memoryUsage int64
+}
+
+type overflowedKeyGroupMetrics struct {
+	topMemoryUsageKeyGroups   []*keyGroupMetrics
+	overflowKeyGroupAggregate keyGroupMetrics
+	keyGroupsCount            int64
+}
+
+type keyGroupsScrapeResult struct {
+	duration          time.Duration
+	metrics           []map[string]*keyGroupMetrics
+	overflowedMetrics []*overflowedKeyGroupMetrics
+}
+
+func (e *Exporter) extractKeyGroupMetrics(ch chan<- prometheus.Metric, c redis.Conn, dbCount int) {
+	allDbKeyGroupMetrics := e.gatherKeyGroupsMetricsForAllDatabases(c, dbCount)
+	if allDbKeyGroupMetrics != nil {
+		for db, dbKeyGroupMetrics := range allDbKeyGroupMetrics.metrics {
+			dbLabel := fmt.Sprintf("db%d", db)
+			registerKeyGroupMetrics := func(metrics *keyGroupMetrics) {
+				e.registerConstMetricGauge(
+					ch,
+					"key_group_count",
+					float64(metrics.count),
+					dbLabel,
+					metrics.keyGroup,
+				)
+				e.registerConstMetricGauge(
+					ch,
+					"key_group_memory_usage_bytes",
+					float64(metrics.memoryUsage),
+					dbLabel,
+					metrics.keyGroup,
+				)
+			}
+			if allDbKeyGroupMetrics.overflowedMetrics[db] != nil {
+				overflowedMetrics := allDbKeyGroupMetrics.overflowedMetrics[db]
+				for _, metrics := range overflowedMetrics.topMemoryUsageKeyGroups {
+					registerKeyGroupMetrics(metrics)
+				}
+				registerKeyGroupMetrics(&overflowedMetrics.overflowKeyGroupAggregate)
+				e.registerConstMetricGauge(ch, "number_of_distinct_key_groups", float64(overflowedMetrics.keyGroupsCount), dbLabel)
+			} else if dbKeyGroupMetrics != nil {
```

Comment:
Not sure I follow - why the if/else ? can't you have both, overflow and dbKeyGroupMetrics, for a db ?

Reply:
Fixed.

Reply:
I'm fine with the check, i'm just wondering if the `else` is needed.